application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java

```diff
@@ -203,11 +209,98 @@ public void updateData(String fromVersion) throws Exception {
                     log.info("Skipping edge events migration");
                 }
                 break;
+            case "3.5.0":
+                log.info("Updating data from version 3.5.0 to 3.5.1 ...");
+                upgradeRuleNodes();
+                break;
             default:
                 throw new RuntimeException("Unable to update data, unsupported fromVersion: " + fromVersion);
         }
     }
 
+    private void upgradeRuleNodes() {
+        try {
+            log.info("Lookup rule nodes to upgrade ...");
+            ArrayList<TbVersionedNode> tbVersionedNodes = getTbVersionedNodes();
+            log.info("Found {} versioned nodes to check for upgrade!", tbVersionedNodes.size());
+            for (TbVersionedNode tbVersionedNode : tbVersionedNodes) {
+                String ruleNodeType = tbVersionedNode.getClass().getName();
+                String ruleNodeTypeForLogs = tbVersionedNode.getClass().getSimpleName();
+                int toVersion = tbVersionedNode.getCurrentVersion();
+                log.info("Going to check for nodes with type: {} to upgrade to version: {}.", ruleNodeTypeForLogs, toVersion);
+                var ruleNodesToUpdate = new PageDataIterable<>(
+                        pageLink ->
+                                ruleChainService.findAllRuleNodesByTypeAndVersionLessThan(
+                                        ruleNodeType,
+                                        toVersion,
+                                        pageLink
+                                ),
+                        1024
+                );
+                if (Iterables.isEmpty(ruleNodesToUpdate)) {
+                    log.info("There are no active nodes with type: {}, or all nodes with this type already set to latest version!", ruleNodeTypeForLogs);
+                } else {
+                    for (RuleNode ruleNode : ruleNodesToUpdate) {
+                        RuleNodeId ruleNodeId = ruleNode.getId();
+                        var oldConfiguration = ruleNode.getConfiguration();
+                        int fromVersion = ruleNode.getConfigurationVersion();
+                        log.info("Going to upgrade rule node with id: {} type: {} fromVersion: {} toVersion: {}",
+                                ruleNodeId,
+                                ruleNodeTypeForLogs,
+                                fromVersion,
+                                toVersion);
+                        try {
+                            TbPair<Boolean, JsonNode> upgradeRuleNodeConfigurationResult = tbVersionedNode.upgrade(fromVersion, oldConfiguration);
+                            if (upgradeRuleNodeConfigurationResult.getFirst()) {
+                                ruleNode.setConfiguration(upgradeRuleNodeConfigurationResult.getSecond());
+                            }
```

Comment:
Add additional logging for upgradeRuleNodeConfigurationResult.getFirst() return false.